File: ExchangeRateData.R
library(httr)
library(readr)
library(dplyr)
library(lubridate)
library(zoo)

# Define the URL
url <- "https://www.federalreserve.gov/datadownload/Output.aspx?rel=H10&series=f838388dca2fd4e8bdfb846f3d2c35df&lastobs=&from=01/01/1971&to=10/09/2024&filetype=csv&label=include&layout=seriescolumn"

# Fetch the data
response <- GET(url)
data <- content(response, "text", encoding = "utf-8")
ERData <- read_csv(data, skip = 6, na = "ND", col_names = c('Date', 'Value'))

# Convert Date to Date format and create a sequence of dates
ERData$Date <- as.Date(ERData$Date)
date_range <- seq(from = min(ERData$Date), to = max(ERData$Date), by = "day")

# Join with the complete date sequence (like reindex in pandas)
ERData <- data.frame(Date = date_range) %>%
  left_join(ERData, by = "Date")

# Linear interpolation for missing values
ERData$Value <- na.approx(ERData$Value)

# Calculate USDCHF and CHFUSD
ERData$USDCHF <- ERData$Value
ERData$CHFUSD <- 1 / ERData$USDCHF

# Calculate LogCHFUSD and LaggedLogCHFUSD
ERData$LogCHFUSD <- log(ERData$CHFUSD)
ERData$LaggedLogCHFUSD <- lag(ERData$LogCHFUSD, 1)

# Calculate LogDifferenceCHFUSD
ERData$LogDifferenceCHFUSD <- ERData$LogCHFUSD - ERData$LaggedLogCHFUSD

write.csv(ERData, file = "ExchangeRateData.csv", row.names = FALSE)

File: Archive/Python/ExchangeRateData.py
import pandas as pd
import requests
from io import StringIO
import numpy as np
url = "https://www.federalreserve.gov/datadownload/Output.aspx?rel=H10&series=f838388dca2fd4e8bdfb846f3d2c35df&lastobs=&from=01/01/1971&to=10/09/2024&filetype=csv&label=include&layout=seriescolumn"
response = requests.get(url)
data = StringIO(response.content.decode('utf-8'))
ERData = pd.read_csv(data, skiprows=5, na_values="ND")
ERData.columns = ['Date', 'Value']
ERData['Date'] = pd.to_datetime(ERData['Date'])
date_range = pd.date_range(start=ERData['Date'].min(), end=ERData['Date'].max(), freq='D')
ERData = ERData.set_index('Date').reindex(date_range).rename_axis('Date').reset_index()
ERData['USDCHF'] = ERData['Value'].interpolate(method='linear')
ERData['CHFUSD'] = 1 / ERData['USDCHF']
ERData['LogCHFUSD'] = np.log(ERData['CHFUSD'])
ERData['LaggedLogCHFUSD'] = ERData['LogCHFUSD'].shift(1)
ERData['LogDifferenceCHFUSD'] = ERData['LogCHFUSD'] - ERData['LaggedLogCHFUSD']
